### PR TITLE
create frontend template

### LINKED ISSUE #113: Set up frontend monorepo
We need to set up the monorepo config for the frontend packages. Technologies to follow those used in the [web-ui-components](https://github.com/DiamondLightSource/web-ui-components) library

### PR DESCRIPTION
**Pull request by kaimsfd**
closes #113 

Added frontend container to docker compose. Created frontend workflow. Created frontend devcontainer. Created frontend gitignore, package.json, src/setupTests.ts, tsconfig.json, yarn.lock.

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -6,7 +6,8 @@
 			"extensions": [
 				"redhat.vscode-yaml",
 				"tsandall.opa",
-        		"ms-kubernetes-tools.vscode-kubernetes-tools"
+        		"ms-kubernetes-tools.vscode-kubernetes-tools",
+				"esbenp.prettier-vscode"
```
Comment (garryod) [inline .devcontainer/devcontainer.json:10]: Probably still don't want `prettier` in our top level dev container as it's only relevant to the frontend

File: .github/workflows/backend-code.yml [hunk 2]
```diff
@@ -47,7 +47,7 @@ jobs:
             --all-targets
             --all-features
             --
-            --deny warnings
```
Comment (garryod) [inline .github/workflows/backend-code.yml:50]: Can remove this change as it's not relevant to the PR

File: .github/workflows/frontend-code.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
+name: Frontend Code
+
+#############################
+# Start the job on all push #
+#############################
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:5]: Can get rid of these comments, the yaml itself should be readable to anyone with a basic understanding of actions

File: .github/workflows/frontend-code.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+on:
+  push:
+    branches-ignore: [main]
+    # Remove the line above to run when pushing to master or main
+  pull_request:
+    branches: [main]
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:11]: This should be the same as our other CI

File: .github/workflows/frontend-code.yml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+jobs:
+  build:
+    # Name the Job
+    name: Lint Code Base
+    # Set the agent to run on
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [18.x]
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:25]: We don't need a matrix if just running against one version

File: .github/workflows/frontend-code.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        # See supported Node.js release schedule at https://nodejs.org/en/about/releases/
+
+    ############################################
+    # Grant status permission for MULTI_STATUS #
+    ############################################
+    permissions:
+      contents: read
+      packages: read
+      statuses: write
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:34]: We don't want to give the workflow any permissions which aren't required

File: .github/workflows/frontend-code.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    ##################
+    steps:
+      ##########################
+      # Checkout the code base #
+      ##########################
+      - name: Checkout Code
+        uses: actions/checkout@v4
+        with:
+          # Full git history is needed to get a proper
+          fetch-depth: 0
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:47]: Do we need all history?

File: .github/workflows/frontend-code.yml [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      ################################
+      # Run Linter against code base #
+      ################################
+      - name: Lint Code Base
+        uses: actions/setup-node@v3
+        env:
+          VALIDATE_ALL_CODEBASE: false
+          DEFAULT_BRANCH: main
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          FILTER_REGEX_EXCLUDE: .*README.md.*
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:58]: Do we need these? The `GITHUB_TOKEN` should already be passed

File: .github/workflows/frontend-code.yml [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        with:
+          node-version: ${{ matrix.node-version }}
+          cache: "npm"
+          cache-dependency-path: frontend/
+
+      - name: yarn install
+        run: yarn install
+        working-directory: frontend/
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:66]: Would it make sense to use [`Boreles/actions-yarn`](https://github.com/Borales/actions-yarn) here?

File: frontend/.devcontainer/devcontainer.json [hunk 10]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "xchemlab frontend",
+    "image": "docker.io/library/node:20.6.0-bookworm",
```
Comment (garryod) [inline frontend/.devcontainer/devcontainer.json:3]: We probably want to use the `docker-compose` here

**Review by garryod — CHANGES_REQUESTED** [review]

File: frontend/package.json [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+      "prettier": "3.0.3",
+      "react-scripts": "5.0.1",
+      "typescript": "5.2.2"
+    },
+    "scripts": {
+      "start": "react-scripts start",
+      "build": "react-scripts build",
+      "test": "react-scripts test",
+      "eject": "react-scripts eject",
+      "format": "prettier --jsx-single-quote --print-width 100 --write src/**/*.{ts,tsx,css}"
```
Comment (garryod) [inline frontend/package.json:26]: Would rather we stuck with the defaults, or defined any changes in a `.prettierrc` file if we want to deviate

File: frontend/tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,14 @@
+{
+    "compilerOptions": {
+        "jsx": "react",
+        "outDir": "./build",
+        "allowJs": true,
```
Comment (garryod) [inline frontend/tsconfig.json:5]: Would rather deny JS, and only allow TS

**Review by garryod — CHANGES_REQUESTED** [review]

File: frontend/tsconfig.json [hunk 13]
```diff
@@ -0,0 +1,14 @@
+{
+    "compilerOptions": {
+        "jsx": "react",
+        "outDir": "./build",
+        "allowJs": false,
+        "target": "es5",
+        "useUnknownInCatchVariables": false,
```
Comment (gfrn) [inline frontend/tsconfig.json:7]: Is there any reason to disable typing errors as `unknown` by default?

File: frontend/package.json [hunk 14]
```diff
@@ -0,0 +1,46 @@
+{
+    "name": "xchemlab-frontend",
+    "version": "0.1.0",
+    "private": true,
+    "dependencies": {
+      "@testing-library/jest-dom": "^5.17.0",
+      "@testing-library/react": "^13.4.0",
+      "@testing-library/user-event": "^13.5.0",
```
Comment (gfrn) [inline frontend/package.json:8]: Testing libraries should be `devDependencies`

File: .github/workflows/frontend-code.yml [hunk 15]
```diff
@@ -0,0 +1,33 @@
+name: Frontend Code
+
+on:
+  push:
+  pull_request:
+
+jobs:
+  build:
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:8]: We should de-duplicate jobs like in our other CI jobs

File: .github/workflows/frontend-code.yml [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+    name: Lint Code Base
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:10]: Didn't realise you could name jobs. That's quite nice

File: .github/workflows/frontend-code.yml [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Code
+        uses: actions/checkout@v4
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:14]: Can we use fully qualified versions for actions?

File: .github/workflows/frontend-code.yml (hunk 16 above)
Comment (garryod) [inline .github/workflows/frontend-code.yml:10]: We don't seem to do any linting in this workflow...

File: frontend/.gitignore [hunk 18]
```diff
@@ -0,0 +1,5 @@
+node_modules/
+
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
```
Comment (garryod) [inline frontend/.gitignore:5]: Could we add a comment describing what each section of ignores is for?

**Review by garryod — CHANGES_REQUESTED** [review]

File: .github/workflows/frontend-code.yml (hunk 15 above)
Comment (garryod) [reply .github/workflows/frontend-code.yml:8]: Probably worth the usual comment on this as it's a kinda whacky `if` statement

File: frontend/package.json [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+    "camelcase": "^6.2.1",
+    "mini-css-extract-plugin": "^2.4.5",
+    "bfj": "^7.0.2",
+    "fs-extra": "^10.0.0",
+    "identity-obj-proxy": "^3.0.0",
+    "prompts": "^2.4.2",
+    "react-app-polyfill": "^3.0.0",
+    "react-dev-utils": "^12.0.1",
+    "react-refresh": "^0.11.0",
+    "tailwindcss": "^3.0.2"
```
Comment (garryod) [inline frontend/package.json:58]: Remove this

File: frontend/package.json [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+    "babel-plugin-named-asset-import": "^0.3.8",
+    "babel-preset-react-app": "^10.0.1",
+    "css-loader": "^6.5.1",
+    "postcss-loader": "^6.2.1",
+    "file-loader": "^6.2.0",
+    "sass-loader": "^12.3.0",
+    "source-map-loader": "^3.0.0",
+    "style-loader": "^3.3.1",
+    "resolve-url-loader": "^4.0.0",
+    "postcss": "^8.4.4",
```
Comment (garryod) [inline frontend/package.json:43]: Remove this

File: frontend/package.json [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+    "dotenv-expand": "^5.1.0",
+    "browserslist": "^4.18.1",
+    "@babel/core": "^7.16.0",
+    "babel-loader": "^8.2.3",
+    "babel-plugin-named-asset-import": "^0.3.8",
+    "babel-preset-react-app": "^10.0.1",
+    "css-loader": "^6.5.1",
+    "postcss-loader": "^6.2.1",
+    "file-loader": "^6.2.0",
+    "sass-loader": "^12.3.0",
```
Comment (garryod) [inline frontend/package.json:39]: Remove this

File: frontend/package.json [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+    "eslint-webpack-plugin": "^4.0.1",
+    "prettier": "3.0.3",
+    "typescript": "5.2.2",
+    "@testing-library/jest-dom": "^5.17.0",
+    "@testing-library/react": "^13.4.0",
+    "@testing-library/user-event": "^13.5.0",
+    "web-vitals": "^2.1.4",
+    "@svgr/webpack": "^5.5.0",
+    "babel-jest": "^27.4.2",
+    "jest": "^27.4.3",
```
Comment (garryod) [inline frontend/package.json:17]: Upgrade this to `29.x.x`

Comment (garryod) [reply frontend/package.json:17]: `yarn upgrade-interactive`

**Review by garryod — CHANGES_REQUESTED** [review]
Can we add an `index.ts` file to `src`?

**Comment by kaimsfd** [comment]
webpack/babel configs https://gitlab.diamond.ac.uk/lims/pato-frontend

File: frontend/config/jest/jest.setup.ts [hunk 23]
```diff
@@ -0,0 +1,28 @@
+import "@testing-library/jest-dom";
```
Comment (gfrn) [inline frontend/config/jest/jest.setup.ts:1]: This file is not needed; consider deleting it

**Review by gfrn — CHANGES_REQUESTED** [review]

File: .github/workflows/eslintrc.yml [hunk 24]
```diff
@@ -0,0 +1,10 @@
+env:
```
Comment (garryod) [inline .github/workflows/eslintrc.yml:1]: Do we need an `eslintrc` in workflows? Seems a bit off to me

File: .github/workflows/frontend-code.yml [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+        uses: borales/actions-yarn@v4
+        with:
+          cmd: test --passWithNoTests
+          dir: "frontend"
+
+      - name: yarn eslint-lint
+        uses: borales/actions-yarn@v4
+        with:
+          cmd: eslint-lint
+          dir: "frontend"
```
Comment (garryod) [inline .github/workflows/frontend-code.yml:39]: Could we move this to a separate `lint` job?

File: frontend/config/env.js [hunk 26]
```diff
@@ -0,0 +1,104 @@
+
+
```
Comment (garryod) [inline frontend/config/env.js:2]: Doubt prettier is a fan of this... Another CI job?

**Review by garryod — CHANGES_REQUESTED** [review]

**Review by garryod — APPROVED** [review]

**Pull request merged**